lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py

```diff
@@ -0,0 +1,235 @@
+        return self.clusters[cluster]
+
+    def get_parent_datacenter(self, obj):
+        """ Walk the parent tree to find the objects datacenter """
+        if isinstance(obj, vim.Datacenter):
+            return obj
+        if obj in self.parent_datacenters:
+            return self.parent_datacenters[obj]
+        datacenter = None
+        while True:
+            if not hasattr(obj, 'parent'):
+                break
+            obj = obj.parent
+            if isinstance(obj, vim.Datacenter):
+                datacenter = obj
+                break
+        self.parent_datacenters[obj] = datacenter
+        return datacenter
+
+
+class PyVmomiHelper(object):
+    def __init__(self, module):
+        if not HAS_PYVMOMI:
+            module.fail_json(msg='pyvmomi module required')
+
+        self.module = module
+        self.params = module.params
+        self.content = connect_to_api(self.module)
+        self.cache = PyVmomiCache(self.content, dc_name=self.params['datacenter'])
+
+    def lookup_datastore(self):
+        datastores = self.cache.get_all_objs(self.content, [vim.Datastore], confine_to_datacenter=True)
+        return datastores
+
+    def lookup_datastore_by_cluster(self):
+        cluster = self.cache.get_cluster(self.params['cluster'])
+        if not cluster:
+            self.module.fail_json(msg='Failed to find cluster "%(cluster)s"' % self.params)
+        c_dc = cluster.datastore
+        return c_dc
+
+
+def main():
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(
+        name=dict(type='str'),
+        datacenter=dict(type='str'),
+        cluster=dict(type='str')
+    )
+    module = AnsibleModule(argument_spec=argument_spec,
+                           required_one_of=[
+                               ['cluster', 'datacenter'],
+                           ],
+                           )
+    result = dict(changed=False)
+
+    pyv = PyVmomiHelper(module)
+
+    if module.params['cluster']:
+        dxs = pyv.lookup_datastore_by_cluster()
+    else:
+        dxs = pyv.lookup_datastore()
+
+    datastores = list()
+    for ds in dxs:
+        summary = ds.summary
+        dds = dict()
+        dds['accessible'] = summary.accessible
+        dds['capacity'] = summary.capacity
+        dds['name'] = summary.name
+        dds['freeSpace'] = summary.freeSpace
+        dds['maintenanceMode'] = summary.maintenanceMode
+        dds['multipleHostAccess'] = summary.multipleHostAccess
+        dds['type'] = summary.type
+        dds['uncommitted'] = summary.uncommitted
+        dds['url'] = summary.url
+        # Calculated values
+        dds['provisioned'] = summary.capacity - summary.freeSpace + summary.uncommitted
+        if module.params['name']:
+            if dds['name'] == module.params['name']:
+                datastores.extend([dds])
+        else:
+            datastores.extend([dds])
+
+    result['datastores'] = datastores
+
+    # VM already exists
```

Comment:
if datastore already exists

Reply:
Thank you for all the comments.  I will work on these and update.